File: apps/core/internal/interfaces/httpserver/projects.go
package httpserver

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"

	projectapp "github.com/gofurry/corvus-studio/apps/core/internal/application/project"
	projectdomain "github.com/gofurry/corvus-studio/apps/core/internal/domain/project"
	"github.com/gofurry/corvus-studio/apps/core/internal/interfaces/httpserver/api"
	"github.com/google/uuid"
	"github.com/labstack/echo/v5"
	"go.uber.org/zap"
)

const maxProjectRequestBody = 1 << 20

type ProjectService interface {
	Create(context.Context, projectapp.CreateCommand) (projectdomain.Project, error)
	Get(context.Context, projectdomain.ID) (projectdomain.Project, error)
	List(context.Context) ([]projectdomain.Project, error)
}

func (server *Server) createProjectHandler(c *echo.Context) error {
	request, err := decodeCreateProjectRequest(c)
	if err != nil {
		return server.writeProjectError(c, err)
	}
	description := ""
	if request.Description != nil {
		description = *request.Description
	}
	entity, err := server.projects.Create(c.Request().Context(), projectapp.CreateCommand{
		Name:        request.Name,
		Description: description,
		Location:    request.Location,
		SteamAppID:  request.SteamAppId,
		Language:    request.Language,
		Stage:       projectdomain.Stage(request.Stage),
	})
	if err != nil {
		return server.writeProjectError(c, err)
	}
	response, err := projectResponse(entity)
	if err != nil {
		return server.writeProjectError(c, err)
	}
	c.Response().Header().Set("Location", "/api/v1/projects/"+entity.ID.String())
	return c.JSON(http.StatusCreated, response)
}

func (server *Server) listProjectsHandler(c *echo.Context) error {
	entities, err := server.projects.List(c.Request().Context())
	if err != nil {
		return server.writeProjectError(c, err)
	}
	items := make([]api.Project, 0, len(entities))
	for _, entity := range entities {
		item, err := projectResponse(entity)
		if err != nil {
			return server.writeProjectError(c, err)
		}
		items = append(items, item)
	}
	return c.JSON(http.StatusOK, api.ProjectList{Items: items})
}

func (server *Server) getProjectHandler(c *echo.Context) error {
	id, err := projectdomain.ParseID(c.Param("project_id"))
	if err != nil {
		return server.writeProjectError(c, err)
	}
	entity, err := server.projects.Get(c.Request().Context(), id)
	if err != nil {
		return server.writeProjectError(c, err)
	}
	response, err := projectResponse(entity)
	if err != nil {
		return server.writeProjectError(c, err)
	}
	return c.JSON(http.StatusOK, response)
}

func decodeCreateProjectRequest(c *echo.Context) (api.CreateProjectRequest, error) {
	request := c.Request()
	request.Body = http.MaxBytesReader(c.Response(), request.Body, maxProjectRequestBody)
	decoder := json.NewDecoder(request.Body)
	decoder.DisallowUnknownFields()

	var body api.CreateProjectRequest
	if err := decoder.Decode(&body); err != nil {
		return api.CreateProjectRequest{}, projectdomain.NewValidationError(
			"request",
			"must contain one valid JSON object",
		)
	}
	var trailing json.RawMessage
	if err := decoder.Decode(&trailing); !errors.Is(err, io.EOF) {
		return api.CreateProjectRequest{}, projectdomain.NewValidationError(
			"request",
			"must not contain trailing JSON",
		)
	}
	return body, nil
}

func projectResponse(entity projectdomain.Project) (api.Project, error) {
	id, err := uuid.Parse(entity.ID.String())
	if err != nil || id.Version() != 7 {
		return api.Project{}, fmt.Errorf("encode Project ID %q: expected UUIDv7", entity.ID)
	}
	return api.Project{
		CreatedAt:   entity.CreatedAt.UTC(),
		Description: entity.Description,
		Id:          id,
		Language:    entity.Language,
		Location:    entity.Location,
		Name:        entity.Name,
		Stage:       api.ProjectStage(entity.Stage),
		Status:      api.ProjectStatus(entity.Status),
		SteamAppId:  cloneInt64(entity.SteamAppID),
		UpdatedAt:   entity.UpdatedAt.UTC(),
	}, nil
}

func (server *Server) writeProjectError(c *echo.Context, err error) error {
	status := http.StatusInternalServerError
	code := api.ErrorCodeInternalError
	message := "Core could not complete the request"
	recoverable := false

	switch {
	case errors.Is(err, projectdomain.ErrValidation):
		status = http.StatusBadRequest
		code = api.ErrorCodeValidationFailed
		message = err.Error()
		recoverable = true
	case errors.Is(err, projectdomain.ErrNotFound):
		status = http.StatusNotFound
		code = api.ErrorCodeProjectNotFound
		message = "Project not found"
		recoverable = true
	case errors.Is(err, projectdomain.ErrLocationConflict):
		status = http.StatusConflict
		code = api.ErrorCodeProjectLocationConflict
		message = "A Project already references this location"
		recoverable = true
	default:
		server.logger.Error("Project request failed", zap.Error(err))
	}

	return c.JSON(status, api.ErrorResponse{Error: api.Error{
		Code:        code,
		Message:     message,
		Recoverable: recoverable,
	}})
}

func cloneInt64(value *int64) *int64 {
	if value == nil {
		return nil
	}
	copy := *value
	return &copy
}

File: apps/web/src/api/projects.ts
import {
  createProject as createProjectRequest,
  getProject as getProjectRequest,
  listProjects as listProjectsRequest,
  type CreateProjectRequest,
  type ErrorResponse,
  type Project,
  type ProjectStage,
  type ProjectStatus,
} from '@corvus-studio/api-client';

export type { CreateProjectRequest, Project, ProjectStage, ProjectStatus };

export class ProjectApiError extends Error {
  readonly code: string;
  readonly recoverable: boolean;

  constructor(error: ErrorResponse['error']) {
    super(error.message);
    this.name = 'ProjectApiError';
    this.code = error.code;
    this.recoverable = error.recoverable;
  }
}

const requestBaseURL = window.location.origin;

export const projectsApi = {
  async list(): Promise<Project[]> {
    try {
      const { data } = await listProjectsRequest({
        baseUrl: requestBaseURL,
        throwOnError: true,
      });
      return data.items;
    } catch (error) {
      throw normalizeProjectError(error);
    }
  },

  async create(input: CreateProjectRequest): Promise<Project> {
    try {
      const { data } = await createProjectRequest({
        baseUrl: requestBaseURL,
        body: input,
        throwOnError: true,
      });
      return data;
    } catch (error) {
      throw normalizeProjectError(error);
    }
  },

  async get(projectId: string): Promise<Project> {
    try {
      const { data } = await getProjectRequest({
        baseUrl: requestBaseURL,
        path: { project_id: projectId },
        throwOnError: true,
      });
      return data;
    } catch (error) {
      throw normalizeProjectError(error);
    }
  },
};

function normalizeProjectError(error: unknown): Error {
  if (isErrorResponse(error)) {
    return new ProjectApiError(error.error);
  }
  if (error instanceof Error) {
    return error;
  }
  return new Error('Corvus Core could not be reached');
}

function isErrorResponse(value: unknown): value is ErrorResponse {
  if (typeof value !== 'object' || value === null || !('error' in value)) {
    return false;
  }
  const error = value.error;
  return (
    typeof error === 'object' &&
    error !== null &&
    'code' in error &&
    typeof error.code === 'string' &&
    'message' in error &&
    typeof error.message === 'string' &&
    'recoverable' in error &&
    typeof error.recoverable === 'boolean'
  );
}
